src/virtio-p9fs.cpp

```diff
@@ -0,0 +1,1828 @@
+            } else {
+                ts[1].tv_sec = 0;
+                ts[1].tv_nsec = UTIME_NOW;
+            }
+        } else {
+            ts[1].tv_sec = 0;
+            ts[1].tv_nsec = UTIME_OMIT;
+        }
+        // Use fd when available, because its path might have been removed while fd still open
+        if (fidp->fd >= 0) {
+            if (futimens(fidp->fd, ts) != 0) {
+                return send_error(msg, tag, host_errno_to_p9(errno));
+            }
+        } else {
+            if (utimensat(AT_FDCWD, fidp->path.c_str(), ts, AT_SYMLINK_NOFOLLOW) != 0) {
+                return send_error(msg, tag, host_errno_to_p9(errno));
+            }
+        }
+        ctime_updated = true;
+    }
+    // Modify change time
+    if ((mask & P9_SETATTR_CTIME) && !ctime_updated) {
+        // Use fd when available, because its path might have been removed while fd still open
+        if (fidp->fd >= 0) {
+            if (fchown(fidp->fd, -1, -1) != 0) {
+                return send_error(msg, tag, host_errno_to_p9(errno));
+            }
+        } else {
+            if (lchown(fidp->path.c_str(), -1, -1) != 0) {
+                return send_error(msg, tag, host_errno_to_p9(errno));
+            }
+        }
+    }
+    // Reply
+    return send_ok(msg, tag, P9_RSETATTR);
+}
+
+bool virtio_p9fs_device::op_readlink(virtq_unserializer &&msg, uint16_t tag) {
+    uint32_t fid{};
+    if (!msg.unpack(&fid)) {
+        return send_error(msg, tag, P9_EPROTO);
+    }
+#ifdef DEBUG_VIRTIO_P9FS
+    (void) fprintf(stderr, "p9fs readlink: tag=%d fid=%d\n", tag, fid);
+#endif
+    // Get the fid state
+    const p9_fid_state *fidp = get_fid_state(fid);
+    if (!fidp) {
+        return send_error(msg, tag, P9_EPROTO);
+    }
+    // Read the link
+    char target[P9_PATH_MAX]{};
+    const ssize_t ret = readlink(fidp->path.c_str(), target, sizeof(target) - 1);
+    if (ret < 0) {
+        return send_error(msg, tag, host_errno_to_p9(errno));
+    }
+    target[ret] = 0;
+    // Reply
+    virtq_serializer out_msg(msg.a, msg.vq, msg.queue_idx, msg.desc_idx, P9_OUT_MSG_OFFSET);
+    if (!out_msg.pack(target)) {
+        return send_error(msg, tag, P9_EPROTO);
+    }
+    return send_reply(std::move(out_msg), tag, P9_RREADLINK);
+}
+
+bool virtio_p9fs_device::op_getattr(virtq_unserializer &&msg, uint16_t tag) {
+    uint32_t fid{};
+    uint64_t mask{};
+    if (!msg.unpack(&fid, &mask)) {
+        return send_error(msg, tag, P9_EPROTO);
+    }
+#ifdef DEBUG_VIRTIO_P9FS
+    (void) fprintf(stderr, "p9fs getattr: tag=%d fid=%d mask=%lx\n", tag, fid, mask);
+#endif
+    // Get the fid state
+    const p9_fid_state *fidp = get_fid_state(fid);
+    if (!fidp) {
+        return send_error(msg, tag, P9_EPROTO);
+    }
+    stat_t st{};
+    // Use fd when available, because its path might have been removed while fd still open
+    if (fidp->fd >= 0) {
+        // Get the attributes
+        if (fstat(fidp->fd, &st) != 0) {
```

Comment:
Is it even worth it to try to get this to work on Windows? Probably a pain, right?

Reply:
I want to at least allow to compile the cartesi machine on Windows / WASM, with some VirtIO being obviously unsupported. This is something I have to do yet with some ifdefs and makefile rules.

Reply:
What does qemu do? (Not saying we should do the same. Just wondering.)

Reply:
QEMU has an abstraction layer for cross platform file operations. For now I am avoiding this problem, and Windows has no support for VirtIO 9PFS, later if there is demand we can add back.